be/src/formats/csv/csv_file_writer.cpp

```diff
@@ -0,0 +1,169 @@
+            }
+
+            if (c != columns.size() - 1) {
+                if (auto st = _output_stream->write(_writer_options->column_terminated_by); !st.ok()) {
+                    return make_ready_future(std::move(st));
+                }
+            }
+        }
+
+        if (auto st = _output_stream->write(_writer_options->line_terminated_by); !st.ok()) {
+            return make_ready_future(std::move(st));
+        }
+    }
+
+    return make_ready_future(Status::OK());
+}
+
+std::future<FileWriter::CommitResult> CSVFileWriter::commit() {
+    auto promise = std::make_shared<std::promise<FileWriter::CommitResult>>();
+    std::future<FileWriter::CommitResult> future = promise->get_future();
+
+    auto task = [output_stream = _output_stream, p = promise, rollback = _rollback_action, row_counter = _num_rows,
+                 location = _location] {
+        // TODO(letian-jiang): check if there are any outstanding io task
+        FileWriter::CommitResult result{
+                .io_status = Status::OK(), .format = ORC, .location = location, .rollback_action = rollback};
+
+        if (auto st = output_stream->finalize(); !st.ok()) {
+            result.io_status.update(st);
+        }
+
+        if (result.io_status.ok()) {
+            result.file_metrics.record_count = row_counter;
+            result.file_metrics.file_size = output_stream->size();
+        }
+
+        p->set_value(result);
+    };
+
+    if (_executors) {
+        bool ok = _executors->try_offer(task);
+        if (!ok) {
+            Status exception = Status::ResourceBusy("submit close file task fails");
+            LOG(WARNING) << exception;
+            promise->set_value(FileWriter::CommitResult{.io_status = exception, .rollback_action = _rollback_action});
+        }
+    } else {
+        task();
+    }
+
+    return future;
+}
+
+CSVFileWriterFactory::CSVFileWriterFactory(std::shared_ptr<FileSystem> fs, const std::string& format,
+                                           const std::map<std::string, std::string>& options,
+                                           const std::vector<std::string>& column_names,
+                                           std::vector<std::unique_ptr<ColumnEvaluator>>&& column_evaluators,
+                                           PriorityThreadPool* executors)
+        : _fs(std::move(fs)),
+          _format(format),
+          _options(options),
+          _column_names(column_names),
+          _column_evaluators(std::move(column_evaluators)),
+          _executors(executors) {}
+
+Status CSVFileWriterFactory::_init() {
+    for (auto& e : _column_evaluators) {
+        RETURN_IF_ERROR(e->init());
+    }
+    _parsed_options = std::make_shared<CSVWriterOptions>();
+    return Status::OK();
+}
+
+StatusOr<std::shared_ptr<FileWriter>> CSVFileWriterFactory::create(const std::string& path) {
+    if (_parsed_options == nullptr) {
+        RETURN_IF_ERROR(_init());
```

Comment:
better to use call_once to avoid concurrent execution

Reply:
Learned today! 

But there is no concurrent invocation by design.

Reply:
Why is there no concurrent? Is this factory not used by many clients?

Reply:
If this factory is not shared, how will it be used?

Reply:
No. A `FileWriterFactory` will be solely owned by a `ConnectorChunkSink`